### DISCUSSION TITLE
Mixing react router dom.

### DISCUSSION THREAD

**Discussion by Ludea** [maxcountryman/axum-login#161] [General]
Hi,

I got a react app.
I bundle it with vitejs
The output files (html and js) are served by axum.
Navigation between react component is handle by react router dom.

Right now I need to protect a private route.
Currently I do this with react, with a grpc call to a rust api to get creds from a db.
Session is not implemented yet.

Can axum-login handle login and session and and navigate between components ( login / private / public components instead of frontend ?

**Comment by maxcountryman** [comment]
> Can axum-login handle login and session and and navigate between components ( login / private / public components instead of frontend ?

Can you expand on this a bit? An example of what you're trying to do  would be helpful.

Generally speaking, yes: `axum-login` can, via `axum`, redirect the user agent however you like.

**Reply by Ludea** [reply]
From React side I do
```
const Routes = (isInstalled: boolean) => {
  const [isLogged, setIsLogged] = useState<boolean>(false);

  useEffect(() => {
    const result = localStorage.getItem("token");
    if (result) {
      setIsLogged(true);
    }
  }, [setIsLogged]);

  const handleConnection = () => {
    setIsLogged(true);
  };

  return [
    {
      element: <AnonymousRoutes isLogged={isLogged} />,
      children: [
        { path: "/login", element: <Login setIsLogged={handleConnection} /> }, //here I get the result of api call
        { path: "/", element: <Presentation /> },
        { path: "/install", element: <Install /> },
        { path: "/forgot", element: <ForgotPassword /> },
      ],
    },
    {
      element: <InstalledRoutes isInstalled={isInstalled} />,
      children: [
        {
          element: <PrivateRoutes isLogged={isLogged} />,
          children: [
            {
              path: "admin",
              element: <Dashboard />,
              children: [
                { index: true, element: <AdminIndex /> },
                { path: "editor", element: <OnlineEditor /> },
                { path: "tables", element: <Tables /> },
              ],
            },
          ],
        },
```

Login component do an api call to a database to retrieve credentials and redirect to admin page.

would like to do the same thing  with axum-login. But I serve only one html file
```pub fn using_serve_dir() -> Router {
    let serve_dir = ServeDir::new("web/dist").fallback(ServeFile::new("web/dist/index.html"));

    Router::new()
        .nest_service("/", serve_dir.clone())
        .fallback_service(serve_dir)
}

pub async fn serve_http(app: Router, port: u16) {
    let addr = SocketAddr::from(([127, 0, 0, 1], port));
    tracing::info!("HTTP server listening on {}", addr);

    axum::Server::bind(&addr)
        .serve(app.layer(TraceLayer::new_for_http()).into_make_service())
        .await
        .unwrap();
}```

**Reply by maxcountryman** [reply]
Given that you want to continue to serve a single HTML file via `axum`, what you could do is use `axum-login` to manage the session via your current API route. In other words, React would still manage routing but you'd lean on `axum-login` to manage the user session. If you wanted `axum` to handle some of the routing, you'd likely need to use React as a proxy in some way.

**Comment by HosMercury** [comment]
if i understand correctly, react send the cookie of the session or an auth bearer token or creds,
then axum know the user then return data or some error , depending on that react does its routing ..
if user tried to access a protected route through react without being authenticated, they would be redirected by react bc of the err response , knowing that they could by reading the source code of react to understand the routes and the design of your app but they cannot access the data from backend.
server can not redirect react directly, but just you can by the response data or err do your front end routing.
